### PR TITLE
Local allocation for primitives

### PR DESCRIPTION
**Pull request by stedolan**
Allows local allocation for various `%primitives`, including Int32, Int64, Nativeint, Float and refs.

Still remaining:
  - [x] local allocation for implicit float allocations (i.e. projection from float array, float records)
  -  ~~local allocation for boxed arguments of C stubs~~ (C primitive support left for another PR)

We could also follow the same approach to allow local allocation of the result of C primitives, by having the C primitive export a second version in the local-returning case. Is this a good idea?

### CODE REVIEW

**Comment by stedolan** [comment]
This patch now depends on #45 - it was hitting the same bug in the testsuite.

**Comment by stedolan** [comment]
This PR is best reviewed commit-by-commit: it contains a nontrivial but reasonably small type system change, followed by a huge but mechanical patch to many primitives.

File: typing/types.ml [hunk 1]
```diff
@@ -671,9 +671,12 @@ module Alloc_mode = struct
 
   let check_const = function
     | Amode m -> Some m
-    | Amodevar v when v.lower = v.upper ->
-       Some v.lower
-    | Amodevar _ -> None
+    | Amodevar v ->
+       begin try compress_vlower v
+       with NotSubmode -> () (* happens during type errors sometimes *) end;
```
Comment (lpw25) [inline typing/types.ml]: I think this might mean that we should adjust `submode_vc` and `submode_vv` to not mutate types until they know they won't fail. It looks like that would be easy, but I might be missing some reason that the mutations are done in the current order.

Another place that might be involved with this -- that we should definitely fix -- is that `Btype.backtrack` does not currently undo mutations of mode variable bounds.

File: typing/ctype.ml [hunk 2]
```diff
@@ -1540,6 +1540,39 @@ let instance_label fixed lbl =
...
+
+let rec instance_prim_locals locals mvar macc finalret ty =
+  match locals, (repr ty).desc with
+  | l :: locals, Tarrow ((lbl,_,mret),arg,ret,commu) ->
+     let marg = prim_mode mvar l in
+     let macc = Alloc_mode.join [marg; mret; macc] in
+     let mret =
+       match locals with
+       | [] -> finalret
+       | _ :: _ -> macc (* curried arrow *)
```
Comment (lpw25) [inline typing/ctype.ml:1556]: This isn't respecting the modes written in the primitive. For example:
```ocaml
external foo : ('a [@local_opt]) -> ('b -> 'c) = "%foo"
```
It seems a pain to support such definitions, but we should probably error on them rather than silently treat them as if they had been written:
```ocaml
external foo : ('a [@local_opt]) -> 'b -> 'c = "%foo"
```

File: typing/typecore.ml [hunk 3]
```diff
@@ -4312,12 +4318,29 @@ and type_ident env expected_mode ?(recarg=Rejected) lid =
...
+  let val_type, kind =
+    match desc.val_kind with
+    | Val_prim prim ->
+       let ty, mode = instance_prim_mode prim (instance desc.val_type) in
+       begin match prim.prim_name with
+       | "%setfield0" ->
+          (* Special case: (:=) is a primitive with both Alloc_local and
+             Alloc_heap versions, but uniquely the Alloc_heap version is
+             cheaper, as it need not use `caml_modify_local`. So in this
+             case we leave it out of [optimise_allocations] *)
```
Comment (lpw25) [inline typing/typecore.ml]: Shouldn't we only register the allocation if the return mode of the primitive is poly?

File: typing/typecore.ml [hunk 4]
```diff
@@ -5834,7 +5860,7 @@ and type_andops env sarg sands expected_ty =
     | { pbop_op = sop; pbop_exp = sexp; pbop_loc = loc; _ } :: rest ->
         if !Clflags.principal then begin_def ();
         let op_path, op_desc = type_binding_op_ident env sop in
-        let op_type = instance op_desc.val_type in
+        let op_type = op_desc.val_type in
```
Comment (lpw25) [inline typing/typecore.ml:5922]: Why is this needed?

File: lambda/translprim.ml [hunk 5]
```diff
@@ -735,22 +737,43 @@ let transl_primitive loc p env ty path =
...
+     let to_alloc_mode = function
+       | Prim_global, _ -> Alloc_heap
+       | Prim_local, _ -> Alloc_local
+       | Prim_poly, _ -> poly_mode in
+     let arg_modes = List.map to_alloc_mode p.prim_native_repr_args in
+     let ret_mode = to_alloc_mode p.prim_native_repr_res in
+     let rec count_nlocal = function
+       | [] -> assert false
+       | [_] ->
+          begin match to_alloc_mode p.prim_native_repr_res with
```
Comment (lpw25) [inline lambda/translprim.ml]: You can use `ret_mode` here

**Review by lpw25 — APPROVED** [review]

File: lambda/translprim.ml [hunk 6]
```diff
@@ -109,270 +108,273 @@ let gen_array_kind =
...
-  match Hashtbl.find primitives_table p.prim_name with
-  | prim -> prim
-  | exception Not_found ->
-      if String.length p.prim_name > 0 && p.prim_name.[0] = '%' then
-        raise(Error(loc, Unknown_builtin_primitive p.prim_name));
-      External p
-
-let lookup_primitive_and_mark_used loc p env path =
-  match lookup_primitive loc p with
+let lookup_primitive loc mode p =
```
Comment (lpw25) [inline lambda/translprim.ml]: This function is assuming that the type of the primitive is the expected one. In particular:
```ocaml
module I64 : sig
   external add : local_ Int64.t -> local_ Int64.t -> local_ Int64.t = "%int64_add"
 end = Int64
```
will type-check and `I64.add` will be typed as returning a local `int64.t`, but it will allocate it on the heap.

I think that this function should probably take the list of concrete argument modes and the concrete return mode instead.

File: typing/types.ml (hunk 1 above)
Comment (stedolan) [reply typing/types.ml]: This change shouldn't be there, I'll remove it. It was meant to be part of another patch to move the mode var stuff to btype.ml and hook it into backtracking.

The reason for the current order is to handle cycles: we don't want to diverge on `a <= b, b <= a`. You could do this by using an auxiliary temporary data structure, but the way I think I'd prefer to structure it is to catch NotSubmode and back out the changes.

File: typing/typecore.ml (hunk 3 above)
Comment (stedolan) [reply typing/typecore.ml]: That is a much better criterion.

File: typing/ctype.ml (hunk 2 above)
Comment (stedolan) [reply typing/ctype.ml:1556]: Good point. This is easy to fix in the arity detection logic in `typedecl.ml`: the `[@curry]` attribute should make arity detection stop.

File: typing/typecore.ml (hunk 4 above)
Comment (stedolan) [reply typing/typecore.ml:5922]: The call to `instance` has moved inside `type_ident` (called from `type_binding_op_ident`), so is no longer needed here.
